sdk/src/Core/Amazon.Runtime/Internal/Util/StreamingUtf8JsonReader.cs

```diff
@@ -0,0 +1,181 @@
+        {
+            int bytesRead = 0;
+            ReadOnlySpan<byte> leftover = buffer.AsSpan().Slice((int)reader.BytesConsumed);
+            if (reader.BytesConsumed < buffer.Length)
+            {
+                // If BytesConsumed is 0 that means that the previous Read failed because the JSON token was too large to fit in the buffer.
+                // In that case we need to resize the buffer and try again to read the JSON token.
+                if (reader.BytesConsumed == 0)
+                {
+                    var resizedBuffer = ArrayPool<byte>.Shared.Rent(Math.Min(int.MaxValue, (buffer.Length * 2)));
+                    Logger.GetLogger(typeof(StreamingUtf8JsonReader)).DebugFormat("Resizing buffer from {0} to {1}", buffer.Length, resizedBuffer.Length);
+                    // copy over the data from the previous read's buffer to the newly resized buffer.
+                    buffer.AsSpan().CopyTo(resizedBuffer);
+                    // return the previous buffer to the pool and set the new buffer to equal the resized buffer.
+                    ArrayPool<byte>.Shared.Return(buffer);
+                    buffer = resizedBuffer;
+                    // fill the new resized buffer with data from the stream. the offset MUST be leftover.Length 
+                    // so we don't overwrite the data that was copied from the previous buffer, and the number of bytes
+                    // we read must be buffer.Length - leftover.Length which is just the second half of the buffer.
+                    bytesRead = FillBuffer(stream, ref buffer, leftover.Length, buffer.Length - leftover.Length);
+                    var resizedSpan = buffer.AsSpan(0, bytesRead + leftover.Length);
+                    reader = new Utf8JsonReader(resizedSpan, isFinalBlock: bytesRead == 0, reader.CurrentState);
+                    // early return since we have a reader
+                    return;
+                }
+                else
+                {
+                    // Move the unprocessed data from the buffer to the start and fill
+                    // remaining space in the buffer with new content from the stream.
+                    leftover.CopyTo(buffer);
+                    bytesRead = FillBuffer(stream, ref buffer, leftover.Length, buffer.Length - leftover.Length);
+                }
+            }
+            else
+            {
+                bytesRead = FillBuffer(stream, ref buffer, 0, buffer.Length);
+            }
+
+            if (bytesRead == 0)
+            {
+                // empty buffer will be returned 
+                reader = new Utf8JsonReader(buffer.AsSpan(0, bytesRead), isFinalBlock: true, reader.CurrentState);
+                return;
+            }
+
+            // passing in 0, bytesRead +leftover.Length is safe here because even if we are on the last buffer block, and leftover is junk, that means bytesRead would've been zero 
```

Comment:
nit: Can you give context for the "passing in 0". Just something like "Passing in 0 for the start position and the span length to (bytesRead + leftover.Length) is safe here ..."